planner/core/expression_rewriter.go

```diff
@@ -1175,6 +1177,34 @@ func (er *expressionRewriter) betweenToExpression(v *ast.BetweenExpr) {
 // Otherwise it should return false to indicate (the caller) that original behavior needs to be performed.
 func (er *expressionRewriter) rewriteFuncCall(v *ast.FuncCallExpr) bool {
 	switch v.FnName.L {
+	// when column is not null, ifnull on such column is not necessary.
+	case ast.Ifnull:
+		if er.isJoin {
+			return false
+		}
+		if len(v.Args) != 2 {
+			er.err = expression.ErrIncorrectParameterCount.GenWithStackByArgs(v.FnName.O)
+			return true
+		}
+		stackLen := len(er.ctxStack)
+		arg1 := er.ctxStack[stackLen-2]
+		arg2 := er.ctxStack[stackLen-1]
+		newArg1, isColumn := arg1.(*expression.Column)
+		if !isColumn {
+			newArg2, isColumn := arg2.(*expression.Column)
+			if !isColumn {
+				return false
+			}
+			newArg1 = newArg2
```

Comment:
this would make `ifnull(0, col)` be substituted to `col`? seems not consistent with https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#function_ifnull